### ISSUE TITLE
Strange Behavor for Reverse and Forward in Contional NF

### ISSUE THREAD

**Issue by www3cam** [VLL-HD/FrEIA#35]
I defined my normalizing flow like this: 

````
def subnet_fc(c_in, c_out):
    return nn.Sequential(nn.Linear(c_in, 32), nn.ReLU(),
                         nn.Linear(32,  c_out))

cond_t = Ff.ConditionNode(1, name='condition')#need to condition the NF on x
nodes_t = [Ff.InputNode(2, name='input')]
'''
for i in range(10):
  nodes.append(Ff.Node(nodes[-1],Fm.InvAutoFC,{},
                      name='FC_{}'.format(i)))#make sure to apply nonlinearity if not already applied (check this)
  
nodes.append(Ff.OutputNode(nodes[-1], name='output'))
'''
for k in range(12):
    nodes_t.append(Ff.Node(nodes_t[-1],
                         Fm.RNVPCouplingBlock,#RNVP has completely wrong output
                         {'subnet_constructor':subnet_fc, 'clamp':2.0},
                         conditions=cond,
                         name=F'coupling_{k}'))
    nodes_t.append(Ff.Node(nodes_t[-1],
                         Fm.PermuteRandom,
                         {'seed':k},
                         name=F'permute_{k}'))
nodes_t.append(Ff.OutputNode(nodes_t[-1], name='output'))

cinn_t = Ff.ReversibleGraphNet(nodes_t + [cond_t])
````

When I perform a pass data forward through the net via a normal base distribution `basedist_t = td.Normal(torch.tensor(0.0), torch.tensor(1.0))`, I get that the forward and the reverse transformations don't agree.  As in: 

````
samp_t2 = basedist_t.sample((20,2)).float()

a_1 = cinn_t(samp_t2, c=w_samp, rev=False)
rev1 = cinn_t(a_1, c=w_samp, rev = True)

print(torch.all(torch.eq(rev1,samp_t2)))
````
Returns False.  Maybe I am doing something incorrectly, but I assume rev = True is the reverse net and rev = True is the forward net.

**Comment by www3cam** [comment]
Actually correction.  It's likely a numerical error because even though: `print(torch.all(torch.eq(rev1,samp_t2)))` returns false, the numbers are the same up to 3 decimal places.  However when I run the optimization routine in this colab notebook: https://colab.research.google.com/drive/1Ng2TshNYmtloeiUyjQpDeCNxhc_R_2_c?usp=sharing and then forward and reverse, the errors are much more noticeable.

**Comment by www3cam** [comment]
Update 2:  I think the problem is when I pickle and unpickle the model.  When I remove this line the problem goes away (but is no longer the correct optimization routine).  

The reason I do this is that I want to backpropagate through the inputs of the reversible net but I don't want to modify the weights of the reversible net in that instance.  It's basically an EM algorithm where the maximization is a minimization of the KL divergence of two distributions and I don't want to modify the weights of the expectation of the model, but I do want to backpropagate through the samples generated by the P distribution.  I'm curious about what my options are to perform this task or if you have a good way to fix the numerical instability that comes with pickling and unpickling repeatedly.

**Comment by ardizzone** [comment]
Hmm, I'm not quite sure what level of numerical difference is acceptable.
In deep neural networks in general, and also in INNs, something on the order of 10^-4 relative error is not surprising (which it sounds like you are getting)

I didn't quite get what you did with the pickling and unpickling (i.e. if you just checked the numerical inversion error of an untrained model), but as the weights get larger (more specifically the jacobian determinant), the numerical errors tend to grow.
Here is an interesting read on the topic: https://arxiv.org/pdf/2006.09347.pdf
Otherwise, it might also have to do with tranferring the model and weights from and back to the gpu.

Your application sounds very interesting, but maybe it would still work with the currect level of numerical accuracy?
The inaccuracies have never given us big trouble in practice.

I will close the issue for now, seeing as it is over a year old and it is likely your work has moved on, but don't hesitate to re-open it if you want us to go into more detail!

**Comment by www3cam** [comment]
Thanks.  Yea I don't even remember what this code was for to be honest.  I might have found pyro code that helped with it.
